### ISSUE TITLE
Alternate Design and Layout

### ISSUE THREAD

**Issue by fpigerre** [2factorauth/twofactorauth#1368]
At the moment, [TwoFactorAuth](https://twofactorauth.org) displays all added websites on the front page inside a number of tables, ordered by category. This creates a number of issues, notably the rendering of the site for mobile users, speed of loading with regards to users who have access to lower bandwidth speeds and scaling the number of websites listed at TwoFactorAuth.org.
Furthermore, adding columns for more data, or sections for peripheral data seems to be quite cumbersome at the moment.

In order to solve these issues, a non-tabular layout could be implemented.
Some suggestions up to this point include the implementation of a carousel-type layout:

> How about a change to something like a carousel where each category would get a separate view?
>    @mxxcon 

as well as the implementation of an icon-based category selection layout:

> And maybe when you go to the site, it shows you a grid of icons. You can choose which category to view, and there's a transition to that category.
>    @jamcat22 

It is possible, however, that moving away from a tabular layout could reduce the simplicity of twofactorauth.org, in turn reducing user engagement.
Would moving away from a tabular layout be beneficial in the long term?

This issue may involve discussion or information relating to pull request #1348.

**Comment by groenroos** [comment]
Personally I'd go with the icon grid as per @jamcat22.

![Grid of icons](http://i.imgur.com/LZmyqWv.jpg)

When you click on one, the data could be appended below, so that the icon-based navigation is retained (no back button ping pong).

![Data](http://i.imgur.com/biXtbAq.jpg)

This would ease navigation on desktop, especially since on the current site, the category nav is in the header, which isn't fixed (so you have to scroll up to even see it).

This should also improve mobile performance, since the whole mess isn't rendered in one go. As for the mobile-friendliness of tables, the major problem is that the majority of the table cells are just empty space. It's quite easy to eliminate this space;

![Mobile layout](http://i.imgur.com/lm5In1m.png)

Not sure how to achieve this in terms of table-based markup. This layout wouldn't look too amazing on a desktop resolution -- I think the table layout works pretty well when you have the space for it.

Also, not really sure if this is part of this conversation, but I would like to suggest adding a local search bar, that would filter the table, much like the [FontAwesome icon search](https://fortawesome.github.io/Font-Awesome/icons/). There's always Ctrl-F, but it'll pick up on any words on the site (in lieu of just site names), and it's a somewhat obscure feature for the average non-techy user.

**Comment by RichJeanes** [comment]
I like the icon grid layout, but it seems kinda bland without the teal nav bar at the top... Though I also like that that color is retained as the selected icon highlight. Nice touch :+1:

**Comment by smholloway** [comment]
The grid layout looks good. Two concerns: 
1. Does this solve our mobile issues? Unless we do on-demand loading then we'll still have a slow-loading site.
2. Given the breadth of the site, I immediately search for services without trying to guess what category the service falls into. I might be weird here, but I can't imagine anyone reading the whole page just to see if their favorite service supports 2FA. When the UI hides elements the user experience around find (ctrl+f) is poor.

**Comment by groenroos** [comment]
One approach to battle the slow loading mobile site could be to load the raw data in JSON format, and render the table by JS on-the-fly when a category is selected. Loading it in JSON would be considerably more svelte in terms of markup, and would avoid a bunch of HTTP requests since none of the images would be loaded before actually necessary. This would make the site load a lot faster, but without the extra delay of handling an AJAX request when a category is selected.

As for searching, this is why I suggested adding in a local search bar of our very own. Ctrl-F indeed won't work well when stuff is hidden, and regardless, it'll search for any words on the page whatsoever (type in "web" to search for WebMD, first result will be the "websites" in the intro paragraph). Also, as mentioned, not all users know to utilize Ctrl-F, and in-page searching is even more hidden and awkward to use on mobile. A built-in search bar on the page would be visible and more easy-to-use for all users on all devices, the search can be more focused, and it can handle hiding and unhiding things as necessary.

**Comment by fpigerre** [comment]
@groenroos, The layout you added where different types of TFA support are listed below a website's name is quite interesting :+1: 
If combined with icons, a similar layout would definitely have the potential to make mobile-browsing of the website a much better experience. Implementing this layout in parallel with a tabular layout for desktop browsing does pose a significant challenge, however.

In the past, @mxxcon has also mentioned the possibility of using [Clusterize.js](https://nexts.github.io/Clusterize.js/) in order to improve the way pages are loaded, as well as  [Jets.js](https://nexts.github.io/Jets.js/) in order to implement a latency-free search bar.
As mentioned previously, the [FontAwesome Icon search bar](https://fortawesome.github.io/Font-Awesome/icons/) seems like an ideal solution with regards to searching.

If the website data is loaded through JSON, is it possible that although the page load/render latency will be decreased, there may be increased latency when selecting a category, that would otherwise not appear if the whole page was loaded when being accessed?

**Comment by mxxcon** [comment]
@groenroos 

> and would avoid a bunch of HTTP requests since none of the images would be loaded before actually necessary.

Current site loads images only when they are in the view, not all of them at once.

**Comment by jamcat22** [comment]
@smholloway @mxxcon brings up a good point. If we don't load images until their in view, the site would be a lot faster with the icon grid. The current speed issue is caused by the fact that to get to another category, you need to scroll past another category, loading all of the images you pass. With this design, only images in the category you would like to view would load.

@groenroos I love the idea of not having a table on mobile devices. Seems simple and smart.

**Comment by mxxcon** [comment]
But do we have technological means to serve different version for mobile?
Can we switch from table-based structure to `div` based?

**Comment by jamcat22** [comment]
What if we just make it responsive?

Have the table turn into the list depending on screen size?

**Comment by mxxcon** [comment]
Isn't it what it is now..to some degree?

**Comment by jamcat22** [comment]
Yes, but that's what #529 was going to solve. The main problem now is that loading times are long, and a long list is a bit hard to navigate on mobile.

**Comment by mxxcon** [comment]
http://www.webpagetest.org/result/151018_FW_3T0/ doesn't look  that bad.
Even on mobile it doesn't look that bad http://www.webpagetest.org/result/151018_YR_3VQ/

**Comment by jamcat22** [comment]
But scrolling though. So hard.

**Comment by groenroos** [comment]
Yup. You'll _really_ want to avoid horizontal scrolling on mobile, especially since mobile browsers often include a feature that horizontal swipe gestures will navigate between pages in history; you make users scroll the table horizontally, you're just begging for them to accidentally navigate away and lose their place.

I think @mxxcon is right in switching to a `div` based layout. It's far easier to make divs look like tables than to make tables look like divs.

**Comment by jamcat22** [comment]
So how do we think we can accomplish the main redesign? I think we can handle searching functionality a bit later.

**Comment by jamcat22** [comment]
Sorry, that was a bit of a dumb question. What I meant to ask was how do we render the tables under each section? I have a demo of the `alternate-layout` branch on my [test site](https://jamcat22gh.gmcd.co). Right now there are only categories.

**Comment by fpigerre** [comment]
Hello!

At the moment, the `alternate-layout` branch contains the most up-to-date implementation of the new design. So far, the category layout is functional (#1480) and filtering capabilities have been added (#1486). The following two GIF files illustrate what can be seen if the site is run from commit 3dff9d3c37f784f80792408d0428d926d0579dd0.

![Filtering Illustrated](https://cloud.githubusercontent.com/assets/3946887/12086174/4d6c1ac4-b313-11e5-970b-3095ca83e1b8.gif) 

Filtering was implemented using the [Jets.js library](https://jets.js.org/). I am not sure whether adding a little more colour to the new layout would be beneficial, and if so, where it should be added.
I am currently working on implementing a better experience for mobile users, however, implementing a mobile layout that does not use a table seems to conflict with the way filtering is currently implemented.

![Opening Categories](https://cloud.githubusercontent.com/assets/3946887/12086199/7733d4fa-b313-11e5-8be1-a67935ea9231.gif)

Thankyou,
psgs :palm_tree:

**Comment by dsoegijono** [comment]
Hi @psgs, it's looking good! I have a question about filtering though, since your gif seems to only cover search for website. Can it also filter by method, which I believe is what #926 is trying to do and discussed in #521?

On search/filter mode, we should also show the table headers and maybe make the `<td>`s fixed width so they're less...messy. Also, what tool did you use to create GIFs?

**Comment by fpigerre** [comment]
Hello @dsoegijono,

To create the two GIF files shown above, I used a tool called [ScreenToGif](https://screentogif.codeplex.com/), which works very well.

The library used to implement filtering, when filtering tables, allows for an option to be set that specifies which columns should be filtered. Unfortunately, due to the way in which twofactorauth.org uses icons instead of text, it seems that filtering by method would be quite inconvenient using [Jets.js](https://jets.js.org/). We may be able to separately create lists of websites using a particular authentication method using JavaScript, or perhaps add a checkbox below the search bar that would allow for only websites supporting a particular method of authentication to be shown.
More information regarding search function callbacks can be found on the options section of the [Jets.js website](https://jets.js.org/).
The options currently used for initializing the website can be seen below:

``` javascript
var jets = new Jets({
  searchTag: '#jets-search',
  contentTag: '.jets-content',
  didSearch: function (searchPhrase) {
    if (searchPhrase == '') {
      $('*.website-table thead').css('display', 'block');
      $('*.website-table').css('display', 'none');
    } else {
      $('*.website-table thead').css('display', 'none');
      $('*.website-table').css('display', 'block');
    }
  },
  columns: [0] // Search by first column only
});
```

:dizzy_face:  I had completely missed the width of the `<td>`s shown on the page when filtering. We should definitely ensure that they are the same width.

**Comment by fpigerre** [comment]
The new layout has now been merged into the `master` branch and is live on [twofactorauth.org](https://twofactorauth.org). Although we are still working on improving the way in which the website can be resized between desktop and mobile layouts, the site should work well for finding out which websites support Two Factor Authentication.
Thankyou very much for the mock-ups you created @groenroos!

**Comment by jamcat22** [comment]
I feel that the problem has been sufficiently solved, so I am closing this issue. Any further suggestions or problems should be addressed in a new issue.
